### PR TITLE
etcdserver: allow 1 learner in cluster

### PR DESCRIPTION
**Pull request by jingyih**
Hard-coded the maximum number of learners to 1.

cc @gyuho @jpbetz @WIZARD-CXY

### CODE REVIEW

**Comment by jingyih** [comment]
Still thinking about how to make this limit reconfigurable for experimenting purpose. Since this is a cluster configuration, seem it is only safe to reconfigure it via quorum write.

**Comment by WIZARD-CXY** [comment]
1 maybe too small for big cluster. We can make it as some function of quorum number.
For example, #maxLearner = #quorum. So it can grow as the cluster grows.
Reconfigure the #maxLearner indeed requires more work. we can leave it afterwards IMHO.

**Review by jpbetz — APPROVED** [review]
lgtm

**Comment by jingyih** [comment]
> 1 maybe too small for big cluster. We can make it as some function of quorum number.
> For example, #maxLearner = #quorum. So it can grow as the cluster grows.
> Reconfigure the #maxLearner indeed requires more work. we can leave it afterwards IMHO.

Currently we do not know how much overhead a learner adds to leader. We do not want adding learner causes cluster to be unavailable. Despite quorum size, there is only 1 leader. If we have more learners, that means more load on leader. See https://github.com/jingyih/etcd/pull/24#issuecomment-486080361 and https://github.com/jingyih/etcd/pull/24#issuecomment-486408857.

File: etcdserver/api/membership/cluster.go [hunk 1]
```diff
@@ -315,6 +316,18 @@ func (c *RaftCluster) ValidateConfigurationChange(cc raftpb.ConfChange) error {
...
 			}
+
+			if confChangeContext.Member.IsLearner { // the new member is a learner
+				numLearners := 0
+				for _, m := range members {
+					if m.IsLearner {
+						numLearners++
+					}
+				}
+				if numLearners+1 > maxLearners {
```
Comment (gyuho) [inline etcdserver/api/membership/cluster.go:327]: Can you help understand why we need `+1`? Or add a comment?

**Comment by gyuho** [comment]
Are we still considering this? Sorry for late review. Added a comment.

File: etcdserver/api/membership/cluster.go (hunk 1 above)
Comment (jingyih) [reply etcdserver/api/membership/cluster.go:327]: The rule is that we can have at most `maxLearners` number of learners. `numLearners` counts the number of existing learners in cluster. `+1` is for the learner member we are adding right now.

**Comment by jingyih** [comment]
> Are we still considering this? Sorry for late review. Added a comment.

Yes we still plan to merge this.

File: etcdserver/api/membership/cluster.go (hunk 1 above)
Comment (gyuho) [reply etcdserver/api/membership/cluster.go:327]: Ah, got it. LGTM>

**Comment by gyuho** [comment]
lgtm, thanks!

**Pull request merged**